File: Searching/c++/interpolationSearch.cpp
#include<bits/stdc++.h> 
using namespace std;
string interpolationSearch(vector<int> a, int l, int r, int num)
{
    r--; 
    while (l <= r && num >= a[l] && num <= a[r])
    {
        // Probing the position with keeping 
        // uniform distribution in mind.
        int pos = l + (((double)(r - l) / (a[r] - a[l])) * (num - a[l]));

        // Condition of target found
        if (a[pos] == num) return("Found");

        // If x is larger, x is in upper part
        if (a[pos] < num) l = pos + 1;

        // If x is smaller, x is in the lower part
        else r = pos - 1;
    }
    return("Not Found");
}

int main()
{
    int n;
    cout << "Enter total number of elements: ";
    cin >> n;
    cout << "Enter Elements";
    int i = n;
    vector<int> a;
    while (i--)
    {
        int t;
        cin >> t;
        a.push_back(t);
    }
    int num;
    cout << "enter number to search";
    cin >> num;
    cout << interpolationSearch(a, 0, n, num);
}

File: Searching/python/interpolationSearch.py
#List should be in Sorted ascending order 

def interpolationSearch(arr, n, x): 
    # Settting indexes of two corners 
    lo = 0
    hi = n 
   
    while (lo <= hi and x >= arr[lo] and x <= arr[hi]): 
        if lo == hi: 
            if arr[lo] == x:  
                return lo; 
            return -1; 

        pos  = lo + int(((float(hi - lo) / 
            ( arr[hi] - arr[lo])) * ( x - arr[lo]))) 
  

        if arr[pos] == x: 
            return pos 
   
        if arr[pos] < x: 
            lo = pos + 1; 
   
        else: 
            hi = pos - 1; 
      
    return -1

print("Enter the elements of the array in a single line: ")
arr = list(map(int,input().split()))
num = int(input("Enter the number you want to search\n"))

index = interpolationSearch(arr, len(arr)-1, num) 

if (index != -1): 
    print ("Element found at index ",index )
else: 
    print ("Element not found")